### ISSUE TITLE
drastic periodic change in epoch times during training, even though nothing else runs on the machine during training.

### ISSUE THREAD

**Issue by Nanex101195** [MIC-DKFZ/nnUNet#1351]
I have a "problem", when training the nnUNet (v1 and v2, same problem) on my NVIDIA DGX A100.

To be honest, I don't know, if it really is a problem, since it appears in a almost perfect periodic way, but as I dug deeper into the code I couldn't find a line, with that I could explain this behavior. 

The setup is:
Nvidia DGXA100 with 4 A100 (40 GB VRAM each)
512 GB RAM 
AMD EPYC 7742 64-Core Processor
Ubuntu 20.04.2 LTS
Python 3.9
nnUNet installation, as stated in your git repository

nnUNet_n_proc_DA = 60, because of 5 modalities to train on (but also tested 32, 40, 50 etc. with the same behavior)

I am the **only one** working on this machine, which makes the behavior even more odd for me.

I train a nnUNet with abdominal MRI with 5 sequences and 8 label classes.
The epoch times range from 43 to 110 seconds. It looks like there's always 4-5 epochs with approx. 40 seconds, 2-3 epochs with approx 110 seconds and then it starts again.
Can you explain this behavior or did you encounter the same "problems" on your machines? 
If not, how could I try to find out, where the problem lies?
I am absolutely clueless on how to debug this..

I attached a progress.png file from the Current training

![progress](https://user-images.githubusercontent.com/45633252/227868064-cba69e76-9794-45be-bdf9-2f63733ce94f.png)

Thank you very much in advance!

**Comment by Nanex101195** [comment]
> Hey,
> 
> very strange indeed. Since I don't observe this behavior I am quite confident that this is not something nnU-Net inherently does, so it is unlikely that there is a bug somewhere that I can simply fix :-/
> 
> So all I can do is make assumptions and pose hypotheses...
> 
> Where is the data located? Is is saved locally on a SSD or do you access it via the network? If the latter is the case then you might have an I/O issue where someone else is bogging down reading of the data. Or the network might be congested with requests.
> 
> Another thing you can do is have a look at the benchmarking documentation (https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/benchmarking.md). It gives some advice on how to locate bottlenecks. You first need to wait for slow epochs to appear and then see how the different components of your system are loaded (CPU, GPU, I/O, RAM) to derive which one is holding back the others.
> 
> What linux kernel do you have? Old kernels can hold back newer hardware
> 
> At some point we had issues with network drive mounting where networking requests would hang for several seconds. This looked differently than your problem but something like that can also be problematic.
> 
> Can you please share your log file from that training as well? It's the txt file located in the folder where the checkpoints are saved.
> 
> Best, Fabian

Thank you for your **very fast** response! 

Well, it still serves as a starting point. So, there must be something wrong with the setup on our DGX A100.
I also encounter other strange issues, when trying to run more than one training at the same time (but this doesn't happen for my own tensorflow implementations of 3DUNets fed by TFRecord files, etc. They scale perfectly, as they should.
The 3D nnUNet on my machine can only run one training, with epoch times of about 43 seconds (as they should be). If I start a second training on a second graphics card, the epoch times get up to 300-400 seconds, which I totally don't understand (even, when nnUNet_DA_preproc is set to something like 20, it just doesn't scale). 
I think these both problems might corellate with each other.

The data is located at the local nvme drive. I tried to have it on the raid nvme and on the system nvme, producing the same behavior. (hdparm showed approx 9GB/s, each. So, shouldn't make a difference, but still tried both)

I already tried to follow the benchmarking documentation and one thing I notice is that during the slow epochs, the "volatile GPU-util" sometimes drops down to 0-20 percent. I don't encounter this during the fast epochs. 
This always happens, when the "main thread?" of the nnUNet is using >180-220 % CPU (seen in "top" in the terminal)) (look at the screenshot below)
When using "top", for "wa" I have never seen another value than zero.
Memory used is just about 50 GB of 512, that looks fine. 
htop still shows free threads and the load seems about no different in the slow epochs, that it does in the slow ones.
In my opinion, Bottleneck issues would be very unlikely, since they appear so periodically and nobody is touching the server, nor anything else is running on it during training and it still happens. Could it help, to set some other environment variables maybe, to limit the number of threads for OMP or something like that? Even if it would slow down things, I'd like to have a controllable environment.


The linux kernel is: 5.4.0-139-generic 

I removed the NAS and the second ethernet connection and deactivated it on the system. same result. 
I will share the log file from the current training, where I encounter the same problems.

The strangest thing in my opinion is that in every test that I did, the first approx. twenty epochs run with the expected epoch duration. After this, the strange things begin to happen.
Looking at the durations alone, it looks a bit like thermal throttling, but doesn't make sense with GPU Temperatures of maximum 66°C. In my tensorflow implementations, I can easily push them to 76°C, without any performance decrease.

I was hoping, nnunetv2 would work better on my machine, but somehow my machine seems to dislike PyTorch in some way. 
As soon, as I saw the nnUNetv2, I created a completely new environment, updated the Nvidia driver and installed everything following exactly your instructions. 

I hope, this response is not too confusing, but maybe you have another idea after all of that? 

thank you so much!

Best regards, 
Florian

**Comment by FabianIsensee** [comment]
Hey, no it's not confusing at all, thanks for the very nicely structured rundown of the problem.
One thing you can try is setting OMP_NUM_THREADS=1, either by `export OMP_NUM_THREADS=1` prior to running the training command or doing `OMP_NUM_THREADS=1 nnUNetv2_train [...]`. 
Can you please also post an htop screenshot from a fast epoch and a slow one? Please make sure to capture the 'training', not the 'validation' phase.
Thanks a lot!
Best,
Fabian

**Comment by FabianIsensee** [comment]
I just ran some training on my local workstation and can confirm that this issue doesn't happen for me 
![progress](https://user-images.githubusercontent.com/11302164/227964753-0fe6bbec-a8cc-404b-aa6c-f0e6ae13f394.png)

**Comment by Nanex101195** [comment]
> Hey, no it's not confusing at all, thanks for the very nicely structured rundown of the problem.
> One thing you can try is setting OMP_NUM_THREADS=1, either by `export OMP_NUM_THREADS=1` prior to running the training command or doing `OMP_NUM_THREADS=1 nnUNetv2_train [...]`.
> Can you please also post an htop screenshot from a fast epoch and a slow one? Please make sure to capture the 'training', not the 'validation' phase.
> Thanks a lot!
> Best,
> Fabian

Hey Fabian, 

I restarted the training several times now and it's always the same behavior...first 10-15 epochs run perfectly fine and then there's this sudden increase in epoch times.
Strange, that you can't encounter the same problem. It's like, one thread just gets off at some point, I don't know.

attached, you can find a screenshot of htop with a fast and a slow epoch, both screenshots were taken approx. 10 seconds, after the print statement of the pseudo dice etc.:

fast: 
![Bildschirm­foto 2023-03-27 um 17 05 44](https://user-images.githubusercontent.com/45633252/227982152-5dbed7d5-c680-4e35-a120-7abefd4d8b93.png)

slow: 
<img width="1419" alt="Bildschirm­foto 2023-03-27 um 16 17 30" src="https://user-images.githubusercontent.com/45633252/227982223-f475fd43-d8df-4090-bc0e-927daedc9e5e.png">


Thank you! 
Best, 
Florian

**Comment by FabianIsensee** [comment]
Hey Florian,
thanks. This all looks fine to me. Have you tried the OMP_NUM_THREADS thing I posted above?
Apart from that I am mostly at a loss. Here are a few more things to try:
- run it inside nvidia ngc pytorch docker
- try different pytorch versions
- have you tried other datasets as well?
Best,
Fabian

**Comment by Nanex101195** [comment]
Hey Fabian, 

sorry for the late reply. I reanimated another machine at our department, with one RTX3090 and it doesn't happen there. (epoch times between 150 and 175s, with a 24 Core AMD CPU).
I think, something with our DGX Station must be messed up.

I'd love to run it inside a Nvidia ngc PyTorch docker, but I know nothing about docker and how to run things. I think this takes some time to learn.

I tried PyTorch 1.13 and PyTorch 2.0.0, python 3.9, python 3.10, different CUDA versions etc.. All with the same results.
Also created a completely new user in ubuntu and installed every packages in a new condo environment. Same result

I experienced this problem with other datasets as well, yes. However, if I set the nnUNet_DA_preproc to something like 24, the epoch times are at about 90-100 seconds and there are not as big jumps. But the graphics cards could do so much more, as we can see when having more DA workers. So, this is just not satisfying at all.

Yesterday I continued a training from epoch 400 and then again, approx. 25 Epochs, everything went fine and then the problems begin. (See the attached "progress.png" file.)

even though, the nnUNet_DA_preproc number is now at 24, I am not able to run two trainings on two different graphics cards at the same time, without receiving epoch times from > 240 seconds...I just don't understand, how this machine handles multithreaded tasks. It seems so messed up.

Also, I performed a nvsm health check that states everything would be healthy.

We think about, completely reinstalling the DGX Station to have a clean system. The station was given to us with a lot of unnecessary consumer things installed onto it, that surely doesn't belong there. Maybe something was messed up from the beginning.

Sorry for bothering you, but do you have access to a DGX Station A100 and could you run a test on it with 50 DA workers or so, to see if the same problems can be reproduced at your site after 30-40 Epochs?

Can pinned memory introduce such problems?

Best,
Florian

**Comment by Nanex101195** [comment]
> Hey Florian, thanks. This all looks fine to me. Have you tried the OMP_NUM_THREADS thing I posted above? Apart from that I am mostly at a loss. Here are a few more things to try:
> 
> * run it inside nvidia ngc pytorch docker
> * try different pytorch versions
> * have you tried other datasets as well?
> 
> Best, Fabian

Hey Fabian,

sorry for the late reply. I reanimated another machine at our department, with one RTX3090 and it doesn't happen there. (epoch times between 150 and 175s, with a 24 Core AMD CPU).
I think, something with our DGX Station must be messed up.

Unfortunately, the OMP_NUM_THREAD=1 didn't change anything of this behavior.

I'd love to run it inside a Nvidia ngc PyTorch docker, but I know nothing about docker and how to run things. I think this takes some time to learn.

I tried PyTorch 1.13 and PyTorch 2.0.0, python 3.9, python 3.10, different CUDA versions etc.. All with the same results.
Also created a completely new user in ubuntu and installed every packages in a new condo environment. Same result

I experienced this problem with other datasets as well, yes. However, if I set the nnUNet_DA_preproc to something like 24, the epoch times are at about 90-100 seconds and there are not as big jumps. But the graphics cards could do so much more, as we can see when having more DA workers. So, this is just not satisfying at all.

Yesterday I continued a training from epoch 400 and then again, approx. 25 Epochs, everything went fine and then the problems begin. (See the attached "progress.png" file.)

even though, the nnUNet_DA_preproc number is now at 24, I am not able to run two trainings on two different graphics cards at the same time, without receiving epoch times from > 240 seconds...I just don't understand, how this machine handles multithreaded tasks. It seems so messed up.

Also, I performed a nvsm health check that states everything would be healthy.

We think about, completely reinstalling the DGX Station to have a clean system. The station was given to us with a lot of unnecessary consumer things installed onto it, that surely doesn't belong there. Maybe something was messed up from the beginning.

Sorry for bothering you, but do you have access to a DGX Station A100 and could you run a test on it with 50 DA workers or so, to see if the same problems can be reproduced at your site after 30-40 Epochs?

Can pinned memory introduce such problems?

Best,
Florian

**Comment by FabianIsensee** [comment]
Hey, I still have no clue. Have you made any progress in the meantime?

**Comment by Nanex101195** [comment]
Sorry for the late reply, but we tried to figure out with Nvidia enterprise support. As a result, we found out there is some hardware malfunctioning that occurs during heavy loads of CPU/GPU and they replace the whole station now.

So....nnUNet helped us to detect malfunctions in our workstation 😄 

Thank you for your help! 

Best regards,
Florian

**Comment by FabianIsensee** [comment]
Hey Florian,
thanks for the update and good luck with the replacement! Let me know if things run smoothly after that :-)
Best,
Fabian

**Comment by sten2lu** [comment]
As this issue seems to already have been solved by Fabian, I will close it now.
